user/libchewing/template.py

```diff
@@ -0,0 +1,47 @@
+pkgname = "libchewing"
+pkgver = "0.9.0"
+pkgrel = 0
+build_style = "cmake"
+# The test fails when run in parallel.
+make_check_env = {"CTEST_PARALLEL_LEVEL": "1"}
+pkgdesc = "Intelligent phonetic input method library"
+maintainer = "xunil-cloud <river_electron@proton.me>"
+license = "LGPL-2.1-only"
+url = "https://chewing.im"
+source = f"https://github.com/chewing/libchewing/releases/download/v{pkgver}/libchewing-{pkgver}.tar.zst"
+sha256 = "58e62cd0649ba3856ffa7c67560c1cfbcbb8713342a533f7700587b51efe84e3"
+tool_flags = {"CFLAGS": ["-DHAVE_NCURSES_H"]}
```

Comment:
this doesn't do anything as cmake already detects this and it's defined in config.h

if it's missing in some file then it's because that file forgot to include config.h and this is the wrong fix

Reply:
looks like the config.h ends up missing all of

```
/* #undef CURSES_HAVE_CURSES_H */
/* #undef CURSES_HAVE_NCURSES_H */
/* #undef CURSES_HAVE_NCURSES_NCURSES_H */
/* #undef CURSES_HAVE_NCURSES_CURSES_H */
```

because nothing actually sets those, then the actual includes are guarded on those

so, you need

```diff
--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -56,6 +56,7 @@
         TEST_HASH_DIR=\"${TEST_BIN_DIR}\"
         TEST_DATA_DIR=\"${TEST_SRC_DIR}/data\"
         TESTDATA=\"${TEST_SRC_DIR}/default-test.txt\"
+        HAVE_CONFIG_H=1
     )
     add_dependencies(${target} data all_static_data)
     add_dependencies(check ${target})
```

for the tests to even try including it (this looks upstreamable), and, as a hack for now

```diff
--- a/cmake/config.h.in
+++ b/cmake/config.h.in
@@ -11,9 +11,7 @@
 #cmakedefine WITH_SQLITE3 1
 
 /* Change cmake curses macro name to autotools curses macro name */
-#ifdef CURSES_HAVE_CURSES_H
 #define HAVE_CURSES_H 1
-#endif
 
 #ifdef CURSES_HAVE_NCURSES_H
 #define HAVE_NCURSES_H 1
```

the reason it doesn't work without this second hack is cmake's findcurses.cmake does not set these values when `set(CURSES_NEED_WIDE true)` is set. why, i have no idea, but they'll just never be defined in that case and then this won't work correctly, so just include the one that works here unconditionally

Reply:
This works:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 0fd9d1a..d68bf0d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -88,6 +88,7 @@ endif()
 include(CheckTypeSize)
 check_type_size(uint16_t UINT16_T)

+include(FindCurses)
 set(CURSES_NEED_WIDE true)
 find_package(Curses)

diff --git a/capi/data/mini.dat b/capi/data/mini.dat
index 435c350..d8aff09 100644
Binary files a/capi/data/mini.dat and b/capi/data/mini.dat differ
```

But If I put `include(FindCurses)` after `set(CURSES_NEED_WIDE true)`, then they'll just never be defined as you said.

Reply:
it works because `include(FindCurses)` is identical to `find_package(Curses)`. so that is equivalent to just removing CURSES_NEED_WIDE, since its ran before being defined

Reply:
(or maybe not "entirely" identical since perhaps find_package does some extra things, but including the same FindCurses.cmake that find_package is already using has that effect here)

Reply:
anyway that doesn't really matter, just use the hack for now